specs/phase1/validator.md

```diff
@@ -0,0 +1,520 @@
+
+Phase 1 adds light client committees and associated responsibilities, discussed [below](#light-client-committee).
+
+### Block proposal
+
+#### Preparing for a `BeaconBlock`
+
+`slot`, `proposer_index`, `parent_root` fields are unchanged.
+
+#### Constructing the `BeaconBlockBody`
+
+`randao_reveal`, `eth1_data`, and `graffiti` are unchanged.
+
+`proposer_slashings`, `deposits`, and `voluntary_exits` are unchanged.
+
+`attester_slashings` and `attestations` operate exactly as in Phase 0, but with new definitations of `AttesterSlashing` and `Attestation`, along with modified validation conditions found in `process_attester_slashing` and `process_attestation`.
+
+##### Custody slashings
+
+Up to `MAX_CUSTODY_SLASHINGS`, [`CustodySlashing`](./custody-game.md#custodyslashing) objects can be included in the `block`. The custody slashings must satisfy the verification conditions found in [custody slashings processing](./custody-game.md#custody-slashings). The validator receives a small "whistleblower" reward for each custody slashing included (THIS IS NOT CURRENTLY THE CASE BUT PROBABLY SHOULD BE).
+
+##### Custody key reveals
+
+Up to `MAX_CUSTODY_KEY_REVEALS`, [`CustodyKeyReveal`](./custody-game.md#custodykeyreveal) objects can be included in the `block`. The custody key reveals must satisfy the verification conditions found in [custody key reveal processing](./custody-game.md#custody-key-reveals). The validator receives a small reward for each custody key reveal included.
+
+##### Early derived secret reveals
+
+Up to `MAX_EARLY_DERIVED_SECRET_REVEALS`, [`EarlyDerivedSecretReveal`](./custody-game.md#earlyderivedsecretreveal) objects can be included in the `block`. The early derived secret reveals must satisfy the verification conditions found in [early derived secret reveal processing](./custody-game.md#custody-key-reveals). The validator receives a small "whistleblower" reward for each early derived secrete reveal included.
+
+##### Shard transitions
+
+Exactly `MAX_SHARDS` [`ShardTransition`](./beacon-chain#shardtransition) objects are included in the block. Default each to an empty `ShardTransition()`. Then for each committee assigned to the slot with an associated `committee_index` and `shard`, set `shard_transitions[shard] = full_transitions[winning_root]` if the committee had enough weight to form a crosslink this slot.
+
+Specifically:
+* Call `shards, winning_roots = get_successful_shard_transitions(state, block.slot, attestations)`
+* Let `full_transitions` be a dictionary mapping from the `shard_transition_root`s found in `attestations` to the corresponding full `ShardTransition`
+* Then for each `shard` and `winning_root` in `zip(shards, winning_roots)` set `shard_transitions[shard] = full_transitions[winning_root]`
+
+```python
+def get_successful_shard_transitions(state: BeaconState,
```

Comment:
Note that since we need to calculate active validator in this function, this `state` has to be at slot `block.slot` to get the correct `current_epoch` value. Is it implied in phase 0 spec that `state` is the lookahead state?

Reply:
It's stated that it must be in the epoch in question to successfully figure out your proposer assignments. I'll make sure this is very clear here